code/Middleware/InitStateMiddleware.php

```diff
@@ -2,46 +2,96 @@
 
 namespace SilverStripe\Subsites\Middleware;
 
+use SilverStripe\Admin\AdminRootController;
 use SilverStripe\Control\HTTPRequest;
 use SilverStripe\Control\Middleware\HTTPMiddleware;
+use SilverStripe\Core\Config\Configurable;
 use SilverStripe\Core\Injector\Injector;
 use SilverStripe\Subsites\Model\Subsite;
 use SilverStripe\Subsites\State\SubsiteState;
 
 class InitStateMiddleware implements HTTPMiddleware
 {
+    use Configurable;
+
+    /**
+     * URL paths that should be considered as admin only, i.e. not frontend
+     *
+     * @config
+     * @var array
+     */
+    private static $admin_url_paths = [
+        'dev/',
+        'graphql/',
+    ];
+
     public function process(HTTPRequest $request, callable $delegate)
     {
+        // Initialise and register the State
         $state = SubsiteState::create();
         Injector::inst()->registerService($state);
 
-        $state->setSubsiteId($this->detectSubsiteId($request));
+        // Detect whether the request was made in the CMS area (or other admin-only areas)
+        $isAdmin = $this->getIsAdmin($request);
+        $state->setUseSessions($isAdmin);
+
+        // Detect the subsite ID
+        $subsiteId = $this->detectSubsiteId($request);
+        $state->setSubsiteId($subsiteId);
+
+        // Persist to the session if using the CMS
+        if ($state->getUseSessions()) {
+            $original = $request->getSession()->get('SubsiteID');
+            $request->getSession()->set('SubsiteID', $subsiteId);
```

Comment:
You should be calling this `Session::set` after the `$delegate($request)`, not immediately after getting the value from the session.